### PR TITLE
`<Dropdown/>` - Add support for Controlled `selectedId`

### PR DESCRIPTION
**Pull request by hepiyellow**
Replaces #2711 

## `upgrade` prop

New `upgrade` prop needs to be added so that Dropdown will support Controlled and Uncontrolled modes (for `selectedId`).
The `upgrade` prop is needed, since we also fix the Uncontrolled behaior.

## Modes
 - Controlled - `selectedId` provided
 - Uncontrolled - `selectedId` is NOT provided. `initiallySelectedId` may be used.

## Tests

I copied ALL the previous tests into a new describe Block called "Upgrade", so that we do regression tests in the NEW APi.

### CODE REVIEW

File: src/Dropdown/Dropdown.js [hunk 1]
```diff
@@ -57,26 +132,74 @@ class Dropdown extends InputWithOptions {
...
+        );
+      }
+      if (props[propName] && props['initiallySelectedId']) {
+        deprecationLog(
+          `'selectedId' and 'initiallySelectedId' cannot both be used at the same time.`,
+        );
+      }
+    },
+  ),
+  initiallySelectedId: validatorWithSideEffect(
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: IMO `initialSelectedId` is a better name, also used in `<BadgeSelect/>` and `<DropdownPopover/>` and defined in the `CONTROLLED_VS_UNCONTROLLED.md` doc.

File: src/Dropdown/Dropdown.js [hunk 2]
```diff
@@ -1,14 +1,55 @@
 import classNames from 'classnames';
 import isUndefined from 'lodash/isUndefined';
+import defaultTo from 'lodash/defaultTo';
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: Why do we have to use it?

File: src/Dropdown/Dropdown.js [hunk 3]
```diff
@@ -1,14 +1,55 @@
 import classNames from 'classnames';
 import isUndefined from 'lodash/isUndefined';
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: Can we remove it? It only check if `value === undefined` (but a better way is probably `typeof value === 'undefined'`).

File: src/Dropdown/Dropdown.js [hunk 4]
```diff
@@ -1,14 +1,55 @@
...
+    const { upgrade, selectedId } = this.props;
+    return upgrade && !isUndefined(selectedId);
+  }
+
+  isControlledSupported() {
+    return this.props.upgrade;
+  }
+
+  static isOptionsEqual(optionsA, optionsB) {
+    return differenceBy(optionsA, optionsB, o => o.id).length === 0;
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: Why? Can't we use `optionA.id === optionB.id`?

File: src/Dropdown/Dropdown.js [hunk 5]
```diff
@@ -1,14 +1,55 @@
...
+      this.deprecatedUpdate(props, { isFirstTime: true });
+    }
+  }
+
+  isSelectedIdControlled() {
+    const { upgrade, selectedId } = this.props;
+    return upgrade && !isUndefined(selectedId);
+  }
+
+  isControlledSupported() {
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: You can remove it. `isSelectedIdControlled` does the same check so it will suffice.

File: src/Dropdown/Dropdown.js [hunk 6]
```diff
@@ -1,14 +1,55 @@
...
+    return this.props.upgrade;
+  }
+
+  static isOptionsEqual(optionsA, optionsB) {
+    return differenceBy(optionsA, optionsB, o => o.id).length === 0;
   }
 
+  getSelectedId() {
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: This can simply be changed to

```js
getSelectedId() {
  return this.isSelectedIdControlled()
    ? this.props.selectedId
    : this.state.selectedId;
}
```

File: src/Dropdown/Dropdown.js [hunk 7]
```diff
@@ -24,7 +65,27 @@ class Dropdown extends InputWithOptions {
     }
   }
 
-  update(props, { isFirstTime }) {
+  /**
+   * Updates the value by the selectedId.
+   * If selectedId is not found in options, then value is NOT changed.
+   */
+  static getNextState(props, selectedId) {
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: Please put all `static` methods at the beginning of the clas.

File: src/Dropdown/Dropdown.js [hunk 8]
```diff
@@ -46,7 +107,21 @@ class Dropdown extends InputWithOptions {
   }
 
   componentWillReceiveProps(nextProps) {
-    this.update(nextProps, { isFirstTime: false });
+    if (this.props.upgrade) {
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: `componentWillReceiveProps` is considered legacy as in React 16. Use `getDerivedStateFromProps` to replace it (using `getDerivedStateFromProps` may also turn `getNextState` redundant).

File: src/Dropdown/Dropdown.js (hunk 7 above)
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: It seems that `getNextState` can be replaced by `getDerivedStateFromProps` (see my comment on `componentWillReceiveProps`).

File: src/Dropdown/Dropdown.js [hunk 9]
```diff
@@ -57,26 +132,74 @@ class Dropdown extends InputWithOptions {
...
+    if (!this.isControlledSupported() || !this.isSelectedIdControlled()) {
+      this.setState({
+        value: this.props.valueParser(option),
+        selectedId: option.id,
+      });
+    }
     super._onSelect(option);
   }
+
+  _onChange(event) {
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: Why is this required?

File: src/Dropdown/Dropdown.js [hunk 10]
```diff
@@ -1,14 +1,55 @@
...
+import defaultTo from 'lodash/defaultTo';
+import differenceBy from 'lodash/differenceBy';
+import { validatorWithSideEffect } from '../utils/propTypes';
+import deprecationLog from '../utils/deprecationLog';
 import InputWithOptions from '../InputWithOptions/InputWithOptions';
 import styles from './Dropdown.scss';
+import PropTypes from 'prop-types';
+
+const NO_SELECTED_ID = null;
+const UPGRADE_PROP_NAME = 'upgrade';
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: Why?

File: src/Dropdown/Dropdown.js [hunk 11]
```diff
@@ -57,26 +132,74 @@ class Dropdown extends InputWithOptions {
...
+          `'selectedId' and 'initiallySelectedId' cannot both be used at the same time.`,
+        );
+      }
+    },
+  ),
+  initiallySelectedId: validatorWithSideEffect(
+    InputWithOptions.propTypes.selectedId,
+    (props, propName) => {
+      if (props[propName] && !props[UPGRADE_PROP_NAME]) {
+        deprecationLog(
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.js]: Maybe it'll be better to return an `Error`.

File: stories/components/Dropdown/Dropdown.story.js [hunk 12]
```diff
@@ -60,6 +60,7 @@ export default {
     options,
     onSelect: option =>
       console.log(`option selected. ${JSON.stringify(option)}.`),
+    upgrade: true,
```
Comment (yardnsm) [inline stories/components/Dropdown/Dropdown.story.js]: I would like to prevent users from toggling this option. Maybe you can move it to `exampleProps` and use something like:

```js
    upgrade: [
		{ label: 'true', value: true },
	],
```

File: src/Dropdown/Dropdown.spec.js [hunk 13]
```diff
@@ -89,6 +89,150 @@ describe('Dropdown', () => {
     expect(driver.isReadOnly()).toBeTruthy();
   });
 
+  describe('Upgrade', () => {
+    const NewDropdown = props => <Dropdown {...props} upgrade />;
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.spec.js]: It seems like we're testing a different component, so I would to the following:

```js
const dropdownProps = {
	upgrade: true,
	options: getOptions(),
};

// In you test
const { inputDriver, dropdownLayoutDriver } = createDriver(
	<Dropdown {...dropdownProps} options={createOverridenOption()} />
);
```

File: src/Dropdown/Dropdown.spec.js [hunk 14]
```diff
@@ -89,6 +89,150 @@ describe('Dropdown', () => {
...
+
+        expect(dropdownLayoutDriver.optionById(0).isSelected()).toBeTruthy();
+        expect(inputDriver.getValue()).toBe('Option 1');
+        rerender(<NewDropdown options={[{ id: 1, value: 'Option 2' }]} />);
+
+        expect(
+          dropdownLayoutDriver
+            .optionDrivers()
+            .some(option => option.isSelected()),
+        ).toBeFalsy();
```
Comment (yardnsm) [inline src/Dropdown/Dropdown.spec.js]: You're only testing a single option, which makes this iteration redundant IMO. Use instead

```js
expect(dropdownLayoutDriver.optionById(1).isSelected()).toBeFalsy();
```

File: src/DropdownLayout/DropdownLayout.driver.js [hunk 15]
```diff
@@ -78,11 +80,18 @@ const dropdownLayoutDriverFactory = ({ element }) => {
...
       doIfOptionExists(position, () => optionAt(position).textContent),
-    /** @deprecated Use optionAtDriver*/
+    /** @deprecated Use optionDriver*/
     optionAt,
     /** Get option driver given an option index */
-    optionDriver: position =>
-      doIfOptionExists(position, () => optionDriver(optionAt(position))),
+    optionDriver: createOptionDriver,
+    /** Get an array of all option drivers */
+    optionDrivers: () => {
```
Comment (yardnsm) [inline src/DropdownLayout/DropdownLayout.driver.js]: Why?

File: src/Dropdown/Dropdown.js (hunk 1 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.js]: ok

File: src/Dropdown/Dropdown.js (hunk 2 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.js]: I like it

File: src/Dropdown/Dropdown.js (hunk 3 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.js]: Let's do a research and add to Contributing Best practices, a white list or/and black-list of `lodash` functions. Until then, "To each his own"

File: src/Dropdown/Dropdown.js (hunk 4 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.js]: This is a comparator of 2 arrays of options, not 2 options.

File: src/Dropdown/Dropdown.js (hunk 2 above)
Comment (argshook) [reply src/Dropdown/Dropdown.js]: please, such "helpers" are useless in these scenarios, just use the language:

`isUndefined` === `typeof thing === 'undefined'`

`defaultTo` === `thing || defaultThing`

File: src/Dropdown/Dropdown.js (hunk 5 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.js]: No. `isSelectedIdControlled` checks if it `selectedId` was passed (hence Controlled).

`isControlledSupported` means that we are in upgrade mode. So we can change the behaviour of Uncontrolled mode as well.

File: src/Dropdown/Dropdown.js (hunk 11 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.js]: Fixed

File: stories/components/Dropdown/Dropdown.story.js (hunk 12 above)
Comment (hepiyellow) [reply stories/components/Dropdown/Dropdown.story.js]: cool

File: src/Dropdown/Dropdown.js (hunk 3 above)
Comment (argshook) [reply src/Dropdown/Dropdown.js]: no, research, contribution list for such things are waste of time. it's as simple as "if problem trivial then use language; otherwise use helper" that's it. `isUndefined` is trivial.

File: src/Dropdown/Dropdown.js [hunk 16]
```diff
@@ -1,14 +1,54 @@
...
 import isUndefined from 'lodash/isUndefined';
+import defaultTo from 'lodash/defaultTo';
+import differenceBy from 'lodash/differenceBy';
+import { allValidators } from '../utils/propTypes';
 import InputWithOptions from '../InputWithOptions/InputWithOptions';
 import styles from './Dropdown.scss';
+import PropTypes from 'prop-types';
+
+const NO_SELECTED_ID = null;
+export const UPGRADE_PROP_NAME = 'upgrade';
```
Comment (argshook) [inline src/Dropdown/Dropdown.js]: i wouldn't do this, just write regular `upgrade` where necessary.

usage of `UPGRADE_PROP_NAME` is inconsistent, so either use it always or never.

also, autodocs wont parse `[UPGRADE_PROP_NAME]` on line 166

---

on a side note, i wish there was no `upgrade` prop at all, but maybe
that's not possible.

i wish, because feature toggles in wix-style-react were never a good
idea, we always suffered one way or another (e.g. `new-icons`, how long will they stay new?)

File: src/Dropdown/Dropdown.js [hunk 17]
```diff
@@ -1,14 +1,54 @@
...
 
 class Dropdown extends InputWithOptions {
   constructor(props) {
     super(props);
-    this.update(props, { isFirstTime: true });
+    if (props.upgrade) {
+      this.state = {
+        value: '',
+        selectedId: NO_SELECTED_ID,
+        ...Dropdown.getNextState(
```
Comment (argshook) [inline src/Dropdown/Dropdown.js]: i would put this to be first, because it seems like you want to override `value` and `selectedId` but if some day those will come from `Dropdown`, there could be issue here

File: src/Dropdown/Dropdown.spec.js (hunk 13 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.spec.js]: In some tests we do `rerender`, we don't use `createDriver` always.
Furthermore, I am going to extract all the old tests to a separate file.

File: src/Dropdown/Dropdown.js (hunk 5 above)
Comment (yardnsm) [reply src/Dropdown/Dropdown.js]: But you always check for `isSelectedIdControlled` **and** ``isControlledSupported`.

File: src/Dropdown/Dropdown.spec.js (hunk 13 above)
Comment (yardnsm) [reply src/Dropdown/Dropdown.spec.js]: > In some tests we do rerender, we don't use createDriver always.

Ok, but I was talking about dropping the `<NewDropdown/>` thing and instead use `<Dropdown {...dropdownProps} />`.

> Furthermore, I am going to extract all the old tests to a separate file.

Why? It's the same component.

File: src/Dropdown/Dropdown.spec.js (hunk 14 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.spec.js]: This iteration is instead of asking `!isAnyOptionSelected`,
I don't want it to be dependent on the ids in the test.
Maybe option id=0 was not removed and it is still selected. So asking about option id=1 might be misleading.

File: src/DropdownLayout/DropdownLayout.driver.js (hunk 15 above)
Comment (hepiyellow) [reply src/DropdownLayout/DropdownLayout.driver.js]: To allow iterating over all options. It would make `optionsLength` deprecated in favor of `optionDrivers().length`.

File: src/Dropdown/Dropdown.js (hunk 16 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.js]: Will change to simple `upgrade`.
About feature toggles. I also wish it, but we are to "heavy" with major versions.
I am trying to figure a way to make it as painless as possible, to do the "cleanup" later.
Like:
- separating the spec tests to different files (e.g `Dropdown.deprecated.spec.js`)
- creating `deprecatedXXXXX` functions, instead of having feature toggles inside functions' logic.

We can also do like the new Button, the has totally 2 separate components.

Would you like to have all "Feature toggles" (API changes) as totaly new components (that is, if not a major version)?

File: src/Dropdown/Dropdown.js (hunk 17 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.js]: I want the value to have a default value of `''`.
And `getNextState` can either return a new value or NOT.

File: src/Dropdown/Dropdown.js (hunk 16 above)
Comment (argshook) [reply src/Dropdown/Dropdown.js]: > Would you like to have all "Feature toggles" (API changes) as totaly new components (that is, if not a major version)?

ideally there should be a way to opt-in to new api without dedicated prop, maybe a new usage of prop
or new value for prop (something like what you did in MultiSelect). But i understand it's not always
possible, just in general saying i don't like feature toggles in components:)

> as totaly new components

that's sometimes possible, for example new `Layout` instead of `Grid`
but that's not the case with Dropdown.

File: src/Dropdown/Dropdown.spec.js (hunk 13 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.spec.js]: I'd rather see it as a pseudo separate component.
This whole `upgrade` feature toggle, is actually a cheap Major version bump.
The idea is to make it as easy as possible to cleanup this mess when we DO bump the major version.
So I separated all tests which we should remove after Major version bump, to a separate file.
And try to use new methods with `deprecatedXXX` names in the component code, to emphasise that we should NOT touch deprecated code, or do fixes in deprecated code.

Furthermore, about your idea to do `<Dropdown {...dropdownProps} />`, I don't want someone to overwrite and pass `upgrade=false`. So if you use the `NewDropdown`, you can NOT make `upgrade=false`.

File: src/Dropdown/Dropdown.js (hunk 5 above)
Comment (hepiyellow) [reply src/Dropdown/Dropdown.js]: But I don't do :  `isSelectedIdControlled` && `isControlledSupported`
